content/blog/scheduling-in-react/index.md

```diff
@@ -0,0 +1,243 @@
+
+## Concurrent React and the Scheduler
+
+_⚠️ Warning: The following APIs are not yet stable and will change. I will do my best to keep this post updated._
+
+To implement a properly scheduled user interface with React, we have to look into two new React features:
+
+- **Concurrent React (also known as Time Slicing).** With the help of the new [Fiber architecture](https://www.youtube.com/watch?v=ZCuYPiUIONs) rewrite that was released with React 16, React can now pause during rendering and yield[^2] to the main thread.
+
+  ➡️ With this feature, we’re able to split long-running rendering tasks into small chunks.
+
+- **Scheduler.** A general purpose cooperative main thread scheduler developed by the React Core team that makes it possible to register callbacks with different priority levels in the browser.
+
+  At the time of writing this article, the priority levels are:
+
+  - `Immediate` for tasks that need to run synchronously.
+  - `UserBlocking` (250ms timeout) for tasks that should run as the result of a user interaction (e.g. a button click).
+  - `Normal` (5s timeout) for rendering non-interactive parts of the screen.
+  - `Low` (10s timeout) for tasks that can be deferred but must still complete eventually (e.g. an analytics notification).
+  - `Idle` (no timeout) for tasks that do not have to run at all. (e.g. hidden offscreen content).
+
+  The timeouts for each priority level are necessary to make sure that lower priority work still runs even if we have so much higher priority work to do that the higher priority work could run continuously. In scheduling algorithms, this problem is referred to as [starvation](<https://en.wikipedia.org/wiki/Starvation_(computer_science)>). The timeouts give us the guarantee that every scheduled task will eventually run. For example, we won’t miss a single analytics notification even if we have ongoing animations in our app.
+
+  ➡️ With this feature, we can schedule tasks with different priorities.
+
+## Scheduling in Action
+
+Let’s see how we can use these features to make an app feel a lot more responsive. To do this, we’ll take a look at [ScheduleTron 3000](https://github.com/philipp-spiess/scheduletron3000), an app I built that allows us to highlight a search term in a list of names. Let’s take a look at the implementation first:
+
+<!-- prettier-ignore -->
+```js
+// The app shows a search box and a list of names. The list is
+// controlled by the searchValue state variable which is updated
+// by the search box.
+function App() {
+  const [searchValue, setSearchValue] = React.useState();
+
+  function handleChange(value) {
+    setSearchValue(value);
+  }
+
+  return (
+    <div>
+      <SearchBox onChange={handleChange} />
+      <NameList searchValue={searchValue} />
+    </div>
+  );
+}
+
+// The search box renders a native HTML input element and keeps
```

Comment:
Just to confirm, it simulates a notification to the servers, or it simulates a notification which is then sent to the servers?

"// The search box renders a native HTML input element and keeps
// it controlled using the inputValue variable. When a new key
// is pressed, it will first update the local inputValue, then
// it will update the App component’s searchValue, and then it will
// simulate an analytics notification to our servers."

Reply:
The first one.